doom/.doom.d/config.org
* UI
#+BEGIN_SRC emacs-lisp
(setq doom-font (font-spec :family "Source Code Pro" :size 14))
(setq doom-theme 'doom-nord-light)
(setq doom-nord-brighter-modeline t)
;; (setq doom-plain-brighter-modeline t)
(setq display-line-numbers-type t)

;; (set-frame-parameter (selected-frame) 'alpha '(90 90))
;; (add-to-list 'default-frame-alist '(alpha 90 90))

(add-to-list 'display-buffer-alist '("*Async Shell Command*" display-buffer-no-window (nil)))

(require 'spaceline-config)
(setq powerline-default-separator 'slant)
(spaceline-spacemacs-theme)
(spaceline-toggle-minor-modes-off)
(setq spaceline-highlight-face-func 'spaceline-highlight-face-evil-state)
(fringe-mode '3)

(remove-hook 'vterm-mode-hook 'hide-mode-line-mode)
(evil-set-initial-state 'vterm-mode 'emacs)
(after! persp-mode
(setq persp-emacsclient-init-frame-behaviour-override "main"))

(setq which-key-idle-delay 0.3)

(setq display-time-day-and-date t)
(display-time)
#+END_SRC

* Org mode
#+BEGIN_SRC emacs-lisp
(setq org-directory "~/sync/")

(map! :map org-mode-map :n "t" 'org-todo)
(map! :map org-agenda-mode-map "C-l" 'org-agenda-log-mode)

(after! org
  (add-to-list 'org-modules 'org-habit)
  (add-to-list 'org-modules 'org-drill)
  (setq org-agenda-files '("~/sync/work" "~/sync/work/projects" "~/sync/work/areas" "~/sync/work/archive"))
  (setq org-agenda-clockreport-parameter-plist '(:link t :maxlevel 2 :fileskip0 t :stepskip0 t))
  (setq org-agenda-skip-scheduled-if-done t)
  (setq org-tags-column -80)
  (setq org-log-done 'time)
  (setq org-startup-folded t)
  (setq org-todo-keywords '((sequence "TODO(t)" "NEXT(n)" "FOLLOWUP(f)" "WAITING(w)" "INACTIVE(i)" "STARTED(s)" "DELEGATED(D@)" "REPEATING(r)" "|" "CANCELLED(c)" "DONE(d)")))
  (org-babel-do-load-languages 'org-babel-load-languages '((ledger . t)))
  (setq org-capture-templates
        '(("P" "Templates for projects")
          ("Pt" "Project-local todo" entry  ; {project-root}/todo.org
           (file+headline +org-capture-project-todo-file "Inbox")
           "* TODO %?\n%i\n%a" :prepend t)
          ("Pn" "Project-local notes" entry  ; {project-root}/notes.org
           (file+headline +org-capture-project-notes-file "Inbox")
           "* %U %?\n%i\n%a" :prepend t)
          ("Pc" "Project-local changelog" entry  ; {project-root}/changelog.org
           (file+headline +org-capture-project-changelog-file "Unreleased")
           "* %U %?\n%i\n%a" :prepend t)
          ("b" "Personal bookmark" entry (file "~/sync/personal.org")
           "* TODO %?")
          ("t" "Work TODO" entry (file+headline "~/sync/work/index.org" "Incoming")
           "* TODO %?")
          ("T" "Work TODO (clocked)" entry (file+headline "~/sync/work/index.org" "Incoming")
           "* STARTED %?" :clock-in t :clock-resume t)
          ("p" "Phone call" entry (file+headline "~/sync/work/index.org" "Incoming")
           "* TODO %? :PHONE:" :clock-in t :clock-resume t))))
#+END_SRC
* Org roam
#+BEGIN_SRC emacs-lisp
(setq org-roam-directory "~/sync/roam")
(setq org-roam-v2-ack t)
#+END_SRC
* Web mode
#+BEGIN_SRC emacs-lisp
(add-to-list 'auto-mode-alist '("\\.leex\\'" . web-mode))
(add-to-list 'auto-mode-alist '("\\.vue\\'" . web-mode))

(defun web-hook ()
  (setq web-mode-markup-indent-offset 2)
  (setq web-mode-css-indent-offset 2)
  (setq web-mode-code-indent-offset 2))

(add-hook 'web-mode-hook 'web-hook)
#+END_SRC

* Elixir mode
#+BEGIN_SRC emacs-lisp
(add-hook 'elixir-mode-hook
          (lambda () (add-hook 'before-save-hook 'elixir-format nil t)))

;; (map! :map elixir-mode-map :nv "m" alchemist-mode-keymap)
#+END_SRC

* Global bindings
#+BEGIN_SRC emacs-lisp
(map!
 "C-c c" 'org-capture
 "s-C" 'kill-buffer-and-window
 "s-<return>" (lambda () (interactive) (start-process-shell-command "st" nil "st"))
 "M-[" 'previous-buffer
 "M-]" 'next-buffer

 :leader
 "--" #'+workspace/display
 "-." #'+workspace/switch-to
 "-`" #'+workspace/other
 "-n" #'+workspace/new
 "-l" #'+workspace/load
 "-s" #'+workspace/save
 "-x" #'+workspace/kill-session
 "-d" #'+workspace/delete
 "-r" #'+workspace/rename
 "-R" #'+workspace/restore-last-session
 "-]" #'+workspace/switch-right
 "-[" #'+workspace/switch-left
 "-1" #'+workspace/switch-to-0
 "-2" #'+workspace/switch-to-1
 "-3" #'+workspace/switch-to-2
 "-4" #'+workspace/switch-to-3
 "-5" #'+workspace/switch-to-4
 "-6" #'+workspace/switch-to-5
 "-7" #'+workspace/switch-to-6
 "-8" #'+workspace/switch-to-7
 "-9" #'+workspace/switch-to-8
 "-0" #'+workspace/switch-to-final

 "," '+ivy/switch-workspace-buffer
 "j" 'dad-joke
 "l" (lambda () (interactive) (start-process-shell-command "slock" nil "slock"))
 "oA" (lambda () (interactive (org-agenda nil "a")))
 "f$" (lambda () (interactive (async-shell-command "miniserve -u .")))
 "$" (lambda () (interactive) (start-process-shell-command "st" nil "st"))
 "%" 'calc
 "!" 'counsel-shell
 "pe" 'projectile-run-vterm
 "ot" '+vterm/here
 "." (lambda () (interactive) (dired "."))
 "oc" 'calendar
 "ob" (lambda () (interactive) (start-process-shell-command "qutebrowser" nil "qutebrowser"))
 "oe" (lambda () (interactive) (start-process-shell-command "qutebrowser https://rockwoodproducts.com/events" nil "qutebrowser https://rockwoodproducts.com/events"))
 "oi" (lambda () (interactive) (start-process-shell-command "inkscape" nil "inkscape"))
 "o~" (lambda () (interactive (dired "~")))
 "op" (lambda () (interactive (dired "~/projects")))
 "ow" (lambda () (interactive) (find-file "~/sync/work/index.org"))
 "os" (lambda () (interactive) (find-file "~/sync/school/index.org"))
 "r" 'org-roam-node-find)
#+END_SRC

* EXWM
#+BEGIN_SRC emacs-lisp
; (require 'exwm)
; (require 'exwm-config)
; (exwm-config-default)

; (add-hook 'exwm-mode-hook #'doom-mark-buffer-as-real-h)

(setq exwm-manage-configurations '(((string= exwm-class-name "xfreerdp") char-mode t)
                                   ((string= exwm-class-name "Qutebrowser") char-mode t)
                                   ((string= exwm-class-name "qutebrowser") char-mode t)
                                   ((string= exwm-class-name "St") char-mode t)
                                   ((string= exwm-class-name "st-256color") char-mode t)
                                   ((string= exwm-class-name "org.inkscape.Inkscape") char-mode t)))

; (setq exwm-input-global-keys
      ; `(
        ; ([?\s-r] . exwm-input-toggle-keyboard)
        ; ([?\M-x] . counsel-M-x)
        ; ([?\s-p] . counsel-shell)
        ; ([?\s-h] . evil-window-left)
        ; ([?\s-k] . evil-window-up)
        ; ([?\s-j] . evil-window-down)
        ; ([?\s-l] . evil-window-right)
        ; ([?\M-\[] . previous-buffer)
        ; ([?\M-\]] . next-buffer)
        ; ([?\M-1] . +workspace/switch-to-0)
        ; ([?\M-2] . +workspace/switch-to-1)
        ; ([?\M-3] . +workspace/switch-to-2)
        ; ([?\M-4] . +workspace/switch-to-3)
        ; ([?\M-5] . +workspace/switch-to-4)
        ; ([?\M-6] . +workspace/switch-to-5)
        ; ([?\M-7] . +workspace/switch-to-6)
        ; ([?\M-8] . +workspace/switch-to-7)
        ; ([?\M-9] . +workspace/switch-to-8)))

; (push ?\C-g exwm-input-prefix-keys)

; (exwm-input-set-key (kbd "M-y") #'my/exwm-counsel-yank-pop)
; (exwm-input-set-key (kbd "M-Y") '+pass/copy-secret)
; (exwm-input-set-key (kbd "M-SPC") doom-leader-map)

(defun counsel-shell (command)
  (interactive
   (list
    (read-shell-command "$ ")))
  (start-process-shell-command command nil command))

; (defun my/exwm-counsel-yank-pop ()
  ; "Same as `counsel-yank-pop' and paste into exwm buffer."
  ; (interactive)
  ; (let ((inhibit-read-only t)
        ; ;; Make sure we send selected yank-pop candidate to
        ; ;; clipboard:
        ; (yank-pop-change-selection t))
    ; (call-interactively #'counsel-yank-pop))
  ; (when (derived-mode-p 'exwm-mode)
    ; ;; https://github.com/ch11ng/exwm/issues/413#issuecomment-386858496
    ; (exwm-input--set-focus (exwm--buffer->id (window-buffer (selected-window))))
    ; (exwm-input--fake-key ?\C-v)))
#+END_SRC
* Dired
#+BEGIN_SRC emacs-lisp
(defun dired-open ()
  (interactive)
  (setq file (dired-get-file-for-visit))
  (setq ext (file-name-extension file))
  (cond ((string= ext "pdf")
          (start-process-shell-command "zathura" nil (concat "zathura " (shell-quote-argument file))))
        ((string= ext "jpg")
          (start-process-shell-command "sxiv" nil (concat "sxiv " (shell-quote-argument file))))
        ((string= ext "jpeg")
          (start-process-shell-command "sxiv" nil (concat "sxiv " (shell-quote-argument file))))
        ((string= ext "png")
          (start-process-shell-command "sxiv" nil (concat "sxiv " (shell-quote-argument file))))
        ((string= ext "dxf")
          (start-process-shell-command "librecad" nil (concat "librecad " (shell-quote-argument file))))
        ((string= ext "html")
          (start-process-shell-command "qutebrowser" nil (concat "qutebrowser " (shell-quote-argument file))))
        ((string= ext "svg")
          (start-process-shell-command "inkscape" nil (concat "inkscape " (shell-quote-argument file))))
        (t (dired-find-file))))

(map! :map dired-mode-map :nv "RET" 'dired-open)
#+END_SRC
* Misc
#+BEGIN_SRC emacs-lisp
(setq user-full-name "Alexander Mielczarek"
      user-mail-address "almielczarek@gmail.com")
(setq +workspaces-on-switch-project-behavior t)

(setq +workspaces-switch-project-function 'counsel-find-file)

(after! ibuffer
  (map! :map ibuffer-mode-map :n "r" #'ibuffer-update))

(global-auto-revert-mode)

(let ((device-specific-config "~/.doom.d/device.el"))
  (when (file-exists-p device-specific-config)
    (load-file device-specific-config)))
#+END_SRC


doom/.doom.d/config.el
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-

;; Place your private configuration here! Remember, you do not need to run 'doom
;; sync' after modifying this file!


;; Some functionality uses this to identify you, e.g. GPG configuration, email
;; clients, file templates and snippets.
(setq user-full-name "Alexander Mielczarek"
      user-mail-address "almielczarek@gmail.com")

;; Doom exposes five (optional) variables for controlling fonts in Doom. Here
;; are the three important ones:
;;
;; + `doom-font'
;; + `doom-variable-pitch-font'
;; + `doom-big-font' -- used for `doom-big-font-mode'; use this for
;;   presentations or streaming.
;;
;; They all accept either a font-spec, font string ("Input Mono-12"), or xlfd
;; font string. You generally only need these two:
(setq doom-font (font-spec :family "Source Code Pro" :size 12))

;; There are two ways to load a theme. Both assume the theme is installed and
;; available. You can either set `doom-theme' or manually load a theme with the
;; `load-theme' function. This is the default:
(setq doom-theme 'doom-nord-light)

;; If you use `org' and don't want your org files in the default location below,
;; change `org-directory'. It must be set before org loads!
(setq org-directory "~/org/")

;; This determines the style of line numbers in effect. If set to `nil', line
;; numbers are disabled. For relative line numbers, set this to `relative'.
(setq display-line-numbers-type t)

;; Here are some additional functions/macros that could help you configure Doom:
;;
;; - `load!' for loading external *.el files relative to this one
;; - `use-package' for configuring packages
;; - `after!' for running code after a package has loaded
;; - `add-load-path!' for adding directories to the `load-path', relative to
;;   this file. Emacs searches the `load-path' when you load packages with
;;   `require' or `use-package'.
;; - `map!' for binding new keys
;;
;; To get information about any of these functions/macros, move the cursor over
;; the highlighted symbol at press 'K' (non-evil users must press 'C-c g k').
;; This will open documentation for it, including demos of how they are used.
;;
;; You can also try 'gd' (or 'C-c g d') to jump to their definition and see how
;; they are implemented.

(add-to-list 'auto-mode-alist '("\\.leex\\'" . web-mode))
(add-to-list 'auto-mode-alist '("\\.vue\\'" . web-mode))

(defun web-hook ()
  (setq web-mode-markup-indent-offset 2)
  (setq web-mode-css-indent-offset 2)
  (setq web-mode-code-indent-offset 2))

(add-hook 'web-mode-hook 'web-hook)

(after! org
  (add-to-list 'org-modules 'org-habit)
  (add-to-list 'org-modules 'org-drill)
  (setq org-agenda-files '("~/org/work.org" "~/org/personal.org"))
  (setq org-tags-column -80)
  (setq org-log-done 'time)
  (setq org-startup-folded t)
  (setq org-todo-keywords '((sequence "TODO(t)" "NEXT(n)" "FOLLOWUP(f)" "WAITING(w)" "INACTIVE(i)" "STARTED(s)" "DELEGATED(D@)" "REPEATING(r)" "|" "CANCELLED(c)" "DONE(d)")))
  (org-babel-do-load-languages 'org-babel-load-languages '((ledger . t)))
  (setq org-capture-templates
        '(("P" "Templates for projects")
          ("Pt" "Project-local todo" entry  ; {project-root}/todo.org
           (file+headline +org-capture-project-todo-file "Inbox")
           "* TODO %?\n%i\n%a" :prepend t)
          ("Pn" "Project-local notes" entry  ; {project-root}/notes.org
           (file+headline +org-capture-project-notes-file "Inbox")
           "* %U %?\n%i\n%a" :prepend t)
          ("Pc" "Project-local changelog" entry  ; {project-root}/changelog.org
           (file+headline +org-capture-project-changelog-file "Unreleased")
           "* %U %?\n%i\n%a" :prepend t)
          ("b" "Personal bookmark" entry (file "~/org/personal.org")
           "* TODO %?")
          ("t" "Work TODO" entry (file+headline "~/org/work.org" "Tasks")
           "* TODO %?")
          ("T" "Work TODO (clocked)" entry (file+headline "~/org/work.org" "Tasks")
           "* DONE %?" :clock-in t :clock-resume t)
          ("p" "Phone call" entry (file+headline "~/org/work.org" "Tasks")
           "* TODO %? :PHONE:" :clock-in t :clock-resume t))))

(defun counsel-shell (command)
  (interactive
   (list
    (read-shell-command "$ ")))
  (start-process-shell-command command nil command))

(evil-set-initial-state 'vterm-mode 'emacs)

(add-hook 'vterm-mode-hook #'doom-modeline-mode)

(map!
 "M-B" '+ivy/switch-buffer
 "C-c c" 'org-capture
 "s-C" 'kill-buffer-and-window
 "s-<return>" '+vterm/here

 :leader
 "oA" (lambda () (interactive (org-agenda nil "a")))
 "f$" (lambda () (interactive (async-shell-command "miniserve -u .")))
 "pe" 'projectile-run-vterm
 "bb" 'switch-to-buffer
 "." (lambda () (interactive) (dired "."))
 "oc" 'calendar
 "o~" (lambda () (interactive (dired "~")))
 "op" (lambda () (interactive (dired "~/projects")))
 "oD" (lambda () (interactive (dired "~/rockwood/development")))
 "od" (lambda () (interactive (dired "~/.dotfiles")))
 "ow" (lambda () (interactive) (find-file "~/org/work.org")))

(map! :map org-mode-map :n "t" 'org-todo)
(map! :map org-agenda-mode-map "C-l" 'org-agenda-log-mode)

(after! persp-mode
  (setq persp-emacsclient-init-frame-behaviour-override "main"))

(add-hook 'elixir-mode-hook
          (lambda () (add-hook 'before-save-hook 'elixir-format nil t)))

(defvar my-org-capture-before-config nil
  "Window configuration before `org-capture'.")

(defadvice org-capture (before save-config activate)
  "Save the window configuration before `org-capture'."
  (setq my-org-capture-before-config (current-window-configuration)))

(add-hook 'org-capture-mode-hook 'delete-other-windows)

(defun my-org-capture-cleanup ()
  "Clean up the frame created while capturing via org-protocol."
  ;; In case we run capture from emacs itself and not an external app,
  ;; we want to restore the old window config
  (when my-org-capture-before-config
    (set-window-configuration my-org-capture-before-config))
  (-when-let ((&alist 'name name) (frame-parameters))
    (when (equal name "org-protocol-capture")
      (delete-frame))))

(add-hook 'org-capture-after-finalize-hook 'my-org-capture-cleanup)

(require 'exwm)
(require 'exwm-config)
(exwm-config-default)

(require 'ivy-posframe)
;; (setq ivy-posframe-display-functions-alist '((t . ivy-posframe-display-at-frame-top-center)))
(setq ivy-posframe-display-functions-alist '((t . ivy-posframe-display-at-frame-center)))
(setq ivy-posframe-parameters '((parent-frame nil)))
(ivy-posframe-mode 1)

(setq exwm-input-global-keys
      `(([?\s-r] . exwm-input-toggle-keyboard) ([?\M-x] . counsel-M-x) ([?\s-p] . counsel-shell) ([?\s-h] . evil-window-left) ([?\s-k] . evil-window-up) ([?\s-j] . evil-window-down) ([?\s-l] . evil-window-right)))

(push ?\C-g exwm-input-prefix-keys)
(push ?\M-1 exwm-input-prefix-keys)
(push ?\M-2 exwm-input-prefix-keys)
(push ?\M-3 exwm-input-prefix-keys)
(push ?\M-4 exwm-input-prefix-keys)
(push ?\M-5 exwm-input-prefix-keys)
(push ?\M-6 exwm-input-prefix-keys)
(push ?\M-7 exwm-input-prefix-keys)
(push ?\M-8 exwm-input-prefix-keys)
(push ?\M-9 exwm-input-prefix-keys)
(push ?\M-B exwm-input-prefix-keys)

(exwm-input-set-key (kbd "M-y") #'my/exwm-counsel-yank-pop)
(exwm-input-set-key (kbd "M-Y") '+pass/copy-secret)
(exwm-input-set-key (kbd "M-SPC") doom-leader-map)

(setq display-time-day-and-date t)
(display-time)
(global-auto-revert-mode)

(defun my/exwm-counsel-yank-pop ()
  "Same as `counsel-yank-pop' and paste into exwm buffer."
  (interactive)
  (let ((inhibit-read-only t)
        ;; Make sure we send selected yank-pop candidate to
        ;; clipboard:
        (yank-pop-change-selection t))
    (call-interactively #'counsel-yank-pop))
  (when (derived-mode-p 'exwm-mode)
    ;; https://github.com/ch11ng/exwm/issues/413#issuecomment-386858496
    (exwm-input--set-focus (exwm--buffer->id (window-buffer (selected-window))))
    (exwm-input--fake-key ?\C-v)))


(let ((device-specific-config "~/.doom.d/device.el"))
  (when (file-exists-p device-specific-config)
    (load-file device-specific-config)))
